rate_limiter.py

```
@@ -82,3 +84,115 @@ def record_success(self, key: str):
+        return True
+
+    def seconds_until_unblocked(self, key: str, now: float = None) -> float:
+        now = now if now is not None else time.time()
+        row = self._conn.execute(
+            "SELECT blocked_until FROM blocks WHERE key = ?", (key,)
+        ).fetchone()
+        if row is None or now >= row[0]:
+            return 0.0
+        return row[0] - now
+
+    def record_failure(self, key: str, now: float = None):
```

Ruff (0.16.1)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)